src/app/components/artist_details/artist_details_model.rs

```diff
@@ -105,11 +109,68 @@ impl PlaylistModel for ArtistDetailsModel {
         )
     }
 
-    fn actions_for(&self, _: String) -> Option<gio::ActionGroup> {
-        None
+    fn actions_for(&self, id: String) -> Option<gio::ActionGroup> {
+        let songs = self.tracks_ref()?;
+        let song = songs.iter().find(|song| song.id == id)?;
+
+        let group = SimpleActionGroup::new();
+
+        let album_id = song.album.id.clone();
+        let view_album = SimpleAction::new("view_album", None);
+        let dispatcher = self.dispatcher.box_clone();
+        view_album.connect_activate(move |_, _| {
+            dispatcher.dispatch(AppAction::ViewAlbum(album_id.clone()));
+        });
+
+        group.add_action(&view_album);
+
+        for (i, artist) in song.artists.iter().enumerate() {
+            let view_artist = SimpleAction::new(&format!("view_artist_{}", i), None);
+            let dispatcher = self.dispatcher.box_clone();
+            let id = artist.id.clone();
+            view_artist.connect_activate(move |_, _| {
+                dispatcher.dispatch(AppAction::ViewArtist(id.clone()));
+            });
+            group.add_action(&view_artist);
+        }
+
+        let track_id = song.id.clone();
+        let copy_uri = SimpleAction::new("copy_uri", None);
+        copy_uri.connect_activate(move |_, _| {
+            let clipboard = Clipboard::get(&SELECTION_CLIPBOARD);
+            clipboard.set_text(&format!("spotify:track:{}", track_id.clone()));
```

Comment:
not sure but track_id or &track_id might be enough for format, we only need to clone it once to move it into the closure

Reply:
> I'm curious, what kind of use do you see for the spotify URI? (spotify:track:etc)

In the official app if you search for a URI it will take you straight to that thing (playlist, artist, album, track): might be nice to add that here.

You can also use them to generate spotify codes (essentially QR codes you can scan in the app) https://www.spotifycodes.com/ (although it might be better to just let users copy codes directly)

Reply:
Yep i think it would, my bad.